oracle/src/main.rs

```diff
@@ -79,55 +188,44 @@ async fn main() -> Result<(), Error> {
     let (key_pair, _) = opts.account_info.get_key_pair()?;
     let signer = PairSigner::<InterBtcRuntime, _>::new(key_pair);
 
-    let interval = Duration::from_millis(opts.interval_ms);
-    let exchange_rate = FixedU128::checked_from_integer(opts.exchange_rate).ok_or(Error::InvalidExchangeRate)?;
-
-    let conversion_factor = FixedU128::checked_from_rational(
-        10_u128.pow(opts.collateral_decimals),
-        10_u128.pow(opts.wrapped_decimals),
-    )
-    .unwrap();
+    let blockstream_url = if let Some(mut url) = opts.blockstream {
+        url.set_path(&format!("{}/fee-estimates", url.path()));
+        Some(url)
+    } else {
+        None
+    };
+
+    // TODO: configurable collateral currencies
+    let coingecko_url = if let Some(mut url) = opts.coingecko {
+        url.set_path(&format!("{}/simple/price", url.path()));
+        url.set_query(Some("ids=bitcoin&vs_currencies=dot"));
+        Some(url)
+    } else {
+        None
+    };
+
+    let conversion_factor =
+        FixedU128::checked_from_rational(10_u128.pow(DOT_DECIMALS), 10_u128.pow(BTC_DECIMALS)).unwrap();
 
     loop {
-        let exchange_rate = if opts.coingecko {
-            match get_exchange_rate_from_coingecko().await {
-                Ok(exchange_rate) => {
-                    // exchange_rate given in BTC/DOT so there is no need to adjust
-                    FixedU128::checked_from_integer(exchange_rate).unwrap()
-                }
-                Err(err) => {
-                    error!("Could not get exchange rate from CoinGecko: {}", err);
-                    sleep(ERR_RETRY_WAIT).await;
-                    continue;
-                }
-            }
-        } else {
-            exchange_rate
-        };
-
-        let exchange_rate = exchange_rate
-            .checked_mul(&conversion_factor)
-            .ok_or(Error::InvalidExchangeRate)?;
-
-        info!(
-            "Setting exchange rate: {} ({})",
-            exchange_rate,
-            chrono::offset::Local::now()
-        );
-
-        let result = InterBtcParachain::from_url_with_retry(
-            &opts.btc_parachain_url.clone(),
-            signer.clone(),
-            Duration::from_millis(opts.connection_timeout_ms),
-        )
-        .await?
-        .set_exchange_rate(exchange_rate)
-        .await;
-
-        if let Err(e) = result {
-            error!("Error: {}", e.to_string());
+        let parachain_rpc =
+            InterBtcParachain::from_url_with_retry(&opts.btc_parachain_url, signer.clone(), opts.connection_timeout_ms)
+                .await?;
+
+        if let Err(err) = try_join!(
```

Comment:
Is there a reason for using `try_join` over `join`? When one call fails, don't we still want to continue executing the second call?

Reply:
It should return on the first `Err`, see: https://docs.rs/tokio/1.10.0/tokio/macro.try_join.html

Reply:
Ah wait, I misread your comment... Yeah we probably do want `join!`.